server/chat/backend/agent/tools/hpa_vpa_card_tool.py

```
@@ -0,0 +1,582 @@
+
+    # Resolve Slack up-front, OUTSIDE the DB block. Both helpers take their own
+    # pool connection internally, so calling them while this function holds one
+    # would nest two checkouts per invocation and can deadlock the pool under
+    # concurrency. Doing it first also fails fast before a row is claimed.
+    slack = _resolve_slack_target(user_id)
+    if slack.get("error"):
+        return json.dumps({"error": slack["error"]})
+
+    try:
+        with db_pool.get_admin_connection() as conn:
+            with conn.cursor() as cur:
```

Ruff (0.16.0)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)